### PR TITLE
Ensure blocks are created by a distinct set of miners. 

### LINKED ISSUE #1675: Ensure blocks are created by a distinct set of miners.
**Issue summary**

There's a check in `verify_blocks` to ensure all blocks are made by different miners, before they are assembled into a valid tipset. It compares each miner address to the first. This is incorrect and will miss the case of any subsequent miners being equal to each other, but not the first, e.g. [A,B,B].

Relevant code:
https://github.com/ChainSafe/forest/blob/46fe3ff57bf1ea634fe325dc9c17755d576fb25e/blockchain/blocks/src/tipset.rs#L227-L230

- [ ] Check what Lotus is doing. Do they have the same mistake? If so, make sure it's actually a mistake and not a misunderstanding of the filecoin specification.
- [ ] Update the check to ensure the miners are all distinct.

**Other information and links**

### PR DESCRIPTION
**Pull request by tyshko5**
**Summary of changes**
Changes introduced in this pull request:
- 

**Reference issue to close (if applicable)**

Closes #1675

**Other information and links**

### CODE REVIEW

File: blockchain/blocks/src/tipset.rs [hunk 1]
```diff
@@ -246,12 +251,16 @@ where
             header.epoch() == first_header.epoch(),
             "epochs are not equal",
         )?;
-        verify(
-            header.miner_address() != first_header.miner_address(),
-            "miner_addresses are not distinct",
-        )?;
+
+        headers_len += 1;
+        headers_set.insert(*header.miner_address());
```
Comment (LesnyRumcajs) [inline blockchain/blocks/src/tipset.rs:256]: we could fail early by checking if insertion succeeded, given it's a set, no?

[the remark this answers is not in the document]
Comment (lemmih) [reply blockchain/blocks/src/tipset.rs:256]: And remember to include the miner address of the `first_header`.

File: blockchain/blocks/src/tipset.rs [hunk 2]
```diff
@@ -429,4 +445,68 @@ mod test {
...
+        let h0 = BlockHeader::builder()
+            .miner_address(Address::new_id(0))
+            .build()
+            .unwrap();
+        let h1 = BlockHeader::builder()
+            .miner_address(Address::new_id(0))
+            .build()
+            .unwrap();
+        assert!(Tipset::new(vec![h0, h1]).unwrap_err() == Error::InvalidTipset("miner_addresses are not distinct".to_string()));
+    }
```
Comment (lemmih) [inline blockchain/blocks/src/tipset.rs:463]: Good. Could you also add a test that would have caught the error in the previous code? Ie, headers that are not distinct but still different from the first header.

File: blockchain/blocks/src/tipset.rs [hunk 3]
```diff
@@ -246,12 +252,19 @@ where
...
+            headers_set.insert(*header.miner_address()),
             "miner_addresses are not distinct",
         )?;
     }
 
+    verify(
+        headers_len == headers_set.len(),
+        "miner_addresses are not distinct",
+    )?;
+
```
Comment (lemmih) [inline blockchain/blocks/src/tipset.rs:267]: This is no longer necessary.

File: blockchain/blocks/src/tipset.rs [hunk 4]
```diff
@@ -429,4 +445,82 @@ mod test {
...
+            .unwrap();
+        assert!(
+            Tipset::new(vec![h0, h1]).unwrap_err()
+                == Error::InvalidTipset("parent cids are not equal".to_string())
+        );
+    }
+
+    #[test]
+    fn ensure_there_are_blocks() {
+        assert!(Tipset::new(vec![]).unwrap_err() == Error::NoBlocks);
```
Comment (lemmih) [inline blockchain/blocks/src/tipset.rs:524]:
```suggestion
        assert_eq!(Tipset::new(vec![]).unwrap_err(), Error::NoBlocks);
```

**Review by lemmih — APPROVED** [review]
I'm happy. It's tempting to use a peekable iterator but that's not necessary.

**Comment by LesnyRumcajs** [comment]
Does Lotus have it right and it was an issue only in Forest? I'm asking, because if it's not the case, probably we should create an issue in Lotus.

**Comment by lemmih** [comment]
> Does Lotus have it right and it was an issue only in Forest? I'm asking, because if it's not the case, probably we should create an issue in Lotus.

I can't seem to find any similar code in Lotus. They don't do any checks when creating a `FullTipSet`. Maybe the checks happen at some other location but grepping gives me nothing.

**Comment by tyshko5** [comment]
> > Does Lotus have it right and it was an issue only in Forest? I'm asking, because if it's not the case, probably we should create an issue in Lotus.
> 
> I can't seem to find any similar code in Lotus. They don't do any checks when creating a `FullTipSet`. Maybe the checks happen at some other location but grepping gives me nothing.

They do it https://github.com/filecoin-project/lotus/blob/e65fae28de2a44947dd24af8c7dafcade29af1a4/chain/store/store.go#L1012-L1015

**Comment by lemmih** [comment]
> > > Does Lotus have it right and it was an issue only in Forest? I'm asking, because if it's not the case, probably we should create an issue in Lotus.
> > 
> > 
> > I can't seem to find any similar code in Lotus. They don't do any checks when creating a `FullTipSet`. Maybe the checks happen at some other location but grepping gives me nothing.
> 
> They do it https://github.com/filecoin-project/lotus/blob/e65fae28de2a44947dd24af8c7dafcade29af1a4/chain/store/store.go#L1012-L1015

Ah, great. Then we don't need to create an issue.

**Comment by LesnyRumcajs** [comment]
> > > Does Lotus have it right and it was an issue only in Forest? I'm asking, because if it's not the case, probably we should create an issue in Lotus.
> > 
> > 
> > I can't seem to find any similar code in Lotus. They don't do any checks when creating a `FullTipSet`. Maybe the checks happen at some other location but grepping gives me nothing.
> 
> They do it https://github.com/filecoin-project/lotus/blob/e65fae28de2a44947dd24af8c7dafcade29af1a4/chain/store/store.go#L1012-L1015

Looks like a bit different behaviour than what we do, no? I'm just double-checking if we are doing the right thing. Here we do the verification later and we fail everything if there's a duplicate. There, from a quick glimpse, they're just grabbing the first one and issuing a warning. If you guys think this is correct then I believe you, but looks suspicious to me. :)

**Comment by LesnyRumcajs** [comment]
> They are issuing a warning each time. What you are saying is that we need to create an issue on Lotus to replace warning with error?

It's rather us that should not fail miserably. Unless it's already the case and this verification fiasco is handled gently.

**Review by LesnyRumcajs — APPROVED** [review]

**Pull request merged**